### PR TITLE
Event handler and emit

### PR DESCRIPTION
**Pull request by iK4tsu**
```md
What changed:

general
 * removed useless **@system** in most functions 

Event:
 * added a **bool handled** property 

genCallbacks
 * now an event can be connected with genCallback functions by passing a delegate [1]
 * added null check on delegate.funcptr at connect and dispach for compatibility with lambda functions [2] 
 * callback delegates now return bool (this is to prevent event propagation if the event has been handled in the
     callback)
 * added an emit function which calls an `onEvent` function (onEvent must be implemented by the user, this cannot
     be done automatically, this function is entirely optional, and is not used internally) [3]

EventDispacher
 * added null check on delegate.funcptr at connect and dispach for compatibility with lambda functions

example-simplesystem
 * updated example to correspond to the latest changes
```
[1]
```d
class Foo
{
    mixin genCallback!(Foo, SomeEvent) onSomeEvent;

public:
    this()
    {
        onSomeEvent.connect(&onSomeEventDefault); // easy bind
    }

protected:
    void onEvent(Event event)
    {
        // dispach here
    }

    bool onSomeEventDefault(Foo sender, in SomeEvent event)
    {
        return true;
    }
}
```
[2]
```d
// using the example above
void main()
{
    Foo foo = new Foo();
    foo.onSomeEvent.connect((Foo sender, in SomeEvent event) {
        // do stuff here
        return true;
    });
}
```
[3]
```d
// imagine there is a MouseClickEvent which takes x, y 
// imagine there is a FocusEvent
class Foo
{
    mixin genCallback!(Foo, MouseClickEvent) onMouseClick;

    // you could do it like this also
    // mixin genCallback!(Foo, MouseClickEvent, int, "x", int, "y") onMouseClick;
    // but the first is prefered most times :)

    mixin genCallback!(Foo, FocusEvent) onFocus;

public:
    this()
    {
        onSomeEvent.connect(&onMouseClickEvent); // easy bind
        onFocus.connect((Foo, FocusEvent event) { import std.stdio: writeln; event.writeln; }); // easy bind
    }

    void click(int x, int y)
    {
        // useful if you don't have an empty ctor but you want to pass the eventin the callback function
        onMouseClick.emit(scoped!MouseClickEvent(x, y));
    }

    void focus()
    {
        // do what focus does
        onFocus.emit(); // you don't need to pass an event it is created by you if it has an empty ctor
    }

protected:
    void onEvent(Event event)
    {
        // dispach here
        auto ed = scoped!EventDispacher(event);
        ed.dispach!MouseClickEvent(&onMouseClick.dispach);
        ed.dispach!FocusEvent(&onFocus.dispach);
    }

    bool onMouseClickEvent(Foo sender, in MouseClickEvent event)
    {
        return true;
    }
}
void main()
{
    Foo foo = new Foo();
    foo.click(4, 12); // calls Foo.onMouseClickEvent
    foo.focus(); // calls the delegate which outputs toString of that event
}

### CODE REVIEW

File: examples/simplesystem/source/events/button/clickedevent.d [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+	mixin basicEventType!ClickedEvent;
+
+public:
+	this(in uint x, in uint y)
+	{
+		this.x = x;
+		this.y = y;
+	}
+
+	const uint x, y;
```
Comment (ljmf00) [inline examples/simplesystem/source/events/button/clickedevent.d:17]: Maybe you want `immutable` here. `unint` is a literal and you never modify the variable.

File: source/deventsystem/eventdispacher.d [hunk 2]
```diff
@@ -59,7 +59,7 @@ public:
 	 */
 	@system
 	void dispach(T : Event)(in void delegate(T) func)
-		in(func.ptr !is null)
+		in(func.ptr !is null || func.funcptr !is null)
```
Comment (ljmf00) [inline source/deventsystem/eventdispacher.d:62]: You may want `&&` operator here, right?

File: source/deventsystem/event.d [hunk 3]
```diff
@@ -167,38 +182,106 @@ mixin template genCallback(T, E : Event, args ...)
...
+		private enum _eventnames = ",event";
 	}
 
 public:
-	pure
-	void connect(void delegate (T, _types) dg)
-		in (dg.ptr !is null)
+	@system pure
+	void connect(bool delegate (T, _types) dg)
+		in(dg.funcptr !is null || dg.ptr !is null)
```
Comment (ljmf00) [inline source/deventsystem/event.d:212]: You may want `&&` operator here, right?

**Review by ljmf00 — CHANGES_REQUESTED** [review]

File: examples/simplesystem/source/events/button/clickedevent.d (hunk 1 above)
Comment (iK4tsu) [reply examples/simplesystem/source/events/button/clickedevent.d:17]: Sure it's just I have nightmares with `immutable`. Also this is just an example, it's the user's choice how he wants to implement his events. `const` is the same as `immutable` in this case.

File: source/deventsystem/eventdispacher.d (hunk 2 above)
Comment (iK4tsu) [reply source/deventsystem/eventdispacher.d:62]: No, you're passing a delegate for example, `func.ptr is null` however `func.funcptr is not null`. If at least one of them is not null, then func is valid.

File: source/deventsystem/event.d (hunk 3 above)
Comment (iK4tsu) [reply source/deventsystem/event.d:212]: ditto

File: examples/simplesystem/source/events/button/clickedevent.d (hunk 1 above)
Comment (ljmf00) [reply examples/simplesystem/source/events/button/clickedevent.d:17]: > Sure it's just I have nightmares with immutable. 

I understand you. I also had some deep nightmares too. `const` is a soft thing, `immutable` is immutable.

Comment (ljmf00) [reply examples/simplesystem/source/events/button/clickedevent.d:17]: > const is the same as immutable in this case

Sure

**Pull request closed without merging**